File: go/test/commit_comment_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/github-sdk/go"
	"github.com/voxgig-sdk/github-sdk/go/core"

	vs "github.com/voxgig-sdk/github-sdk/go/utility/struct"
)

func TestCommitCommentEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.CommitComment(nil)
		if ent == nil {
			t.Fatal("expected non-nil CommitCommentEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"commit_comment": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.CommitComment(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.SharedConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.CommitComment(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := commit_commentBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "commit_comment." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set GITHUB_TEST_COMMIT_COMMENT_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		commitCommentRef01Ent := client.CommitComment(nil)
		commitCommentRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "commit_comment"}), "commit_comment_ref01"))
		commitCommentRef01Data["commit_sha"] = setup.idmap["commit_sha01"]
		commitCommentRef01Data["owner"] = setup.idmap["owner01"]
		commitCommentRef01Data["repo"] = setup.idmap["repo01"]

		commitCommentRef01DataResult, err := commitCommentRef01Ent.Create(commitCommentRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		commitCommentRef01Data = core.ToMapAny(entityData(commitCommentRef01DataResult))
		if commitCommentRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if commitCommentRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		commitCommentRef01Match := map[string]any{
			"owner": setup.idmap["owner01"],
			"repo": setup.idmap["repo01"],
		}

		commitCommentRef01ListResult, err := commitCommentRef01Ent.List(commitCommentRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		commitCommentRef01List, commitCommentRef01ListOk := commitCommentRef01ListResult.([]any)
		if !commitCommentRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", commitCommentRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(commitCommentRef01List), map[string]any{"id": commitCommentRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		commitCommentRef01DataUp0Up := map[string]any{
			"id": commitCommentRef01Data["id"],
			"owner": setup.idmap["owner"],
			"repo": setup.idmap["repo"],
		}

		commitCommentRef01MarkdefUp0Name := "author_association"
		commitCommentRef01MarkdefUp0Value := fmt.Sprintf("Mark01-commit_comment_ref01_%d", setup.now)
		commitCommentRef01DataUp0Up[commitCommentRef01MarkdefUp0Name] = commitCommentRef01MarkdefUp0Value

		commitCommentRef01ResdataUp0Result, err := commitCommentRef01Ent.Update(commitCommentRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		commitCommentRef01ResdataUp0 := core.ToMapAny(entityData(commitCommentRef01ResdataUp0Result))
		if commitCommentRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if commitCommentRef01ResdataUp0["id"] != commitCommentRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if commitCommentRef01ResdataUp0[commitCommentRef01MarkdefUp0Name] != commitCommentRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", commitCommentRef01MarkdefUp0Name, commitCommentRef01ResdataUp0[commitCommentRef01MarkdefUp0Name])
		}

		// LOAD
		commitCommentRef01MatchDt0 := map[string]any{
			"id": commitCommentRef01Data["id"],
		}
		commitCommentRef01DataDt0Loaded, err := commitCommentRef01Ent.Load(commitCommentRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		commitCommentRef01DataDt0LoadResult := core.ToMapAny(entityData(commitCommentRef01DataDt0Loaded))
		if commitCommentRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if commitCommentRef01DataDt0LoadResult["id"] != commitCommentRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func commit_commentBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "commit_comment", "CommitCommentTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read commit_comment test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse commit_comment test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"commit_comment01", "commit_comment02", "commit_comment03", "repo01", "repo02", "repo03", "commit01", "commit02", "commit03", "commit_sha01", "owner01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("GITHUB_TEST_COMMIT_COMMENT_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"GITHUB_TEST_COMMIT_COMMENT_ENTID": idmap,
		"GITHUB_TEST_LIVE":      "FALSE",
		"GITHUB_TEST_EXPLAIN":   "FALSE",
		"GITHUB_APIKEY":         "",
	})

	idmapResolved := core.ToMapAny(env["GITHUB_TEST_COMMIT_COMMENT_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add owner alias for update test.
	if idmapResolved["owner"] == nil {
		idmapResolved["owner"] = idmapResolved["owner01"]
	}
	// Add repo alias for update test.
	if idmapResolved["repo"] == nil {
		idmapResolved["repo"] = idmapResolved["repo01"]
	}

	if env["GITHUB_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["GITHUB_APIKEY"],
			},
			extraOpts,
		})
		client = sdk.NewGithubSDK(core.ToMapAny(mergedOpts))
	}

	live := env["GITHUB_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["GITHUB_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: php/test/CommitCommentEntityTest.php
<?php
declare(strict_types=1);

// CommitComment entity test

require_once __DIR__ . '/../github_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class CommitCommentEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = GithubSDK::test(null, null);
        $ent = $testsdk->CommitComment(null);
        $this->assertNotNull($ent);
    }

    // Feature #4: the entity stream(action, ...) method runs the op pipeline
    // and yields result items. With the streaming feature active it yields the
    // feature's incremental output; otherwise it falls back to the materialised
    // list so stream always yields.
    public function test_stream(): void
    {
        $seed = [
            "entity" => [
                "commit_comment" => [
                    "s1" => ["id" => "s1"],
                    "s2" => ["id" => "s2"],
                    "s3" => ["id" => "s3"],
                ],
            ],
        ];

        // Fallback: streaming inactive -> yields the materialised list items.
        $base = GithubSDK::test($seed, null);
        $seen = iterator_to_array($base->CommitComment(null)->stream("list", null, null), false);
        $this->assertCount(3, $seen);

        // Inbound: streaming active -> yields each item from the feature.
        $cfg = GithubConfig::shared_config();
        if (isset($cfg["feature"]) && is_array($cfg["feature"]) && isset($cfg["feature"]["streaming"])) {
            $sdk = GithubSDK::test($seed, ["feature" => ["streaming" => ["active" => true]]]);
            $got = [];
            foreach ($sdk->CommitComment(null)->stream("list", null, null) as $item) {
                if (is_array($item) && array_is_list($item)) {
                    foreach ($item as $sub) {
                        $got[] = $sub;
                    }
                } else {
                    $got[] = $item;
                }
            }
            $this->assertCount(3, $got);
        }
    }

    public function test_basic_flow(): void
    {
        $setup = commit_comment_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create", "list", "update", "load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "commit_comment." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set GITHUB_TEST_COMMIT_COMMENT_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $commit_comment_ref01_ent = $client->CommitComment(null);
        $commit_comment_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.commit_comment"), "commit_comment_ref01"));
        $commit_comment_ref01_data["commit_sha"] = $setup["idmap"]["commit_sha01"];
        $commit_comment_ref01_data["owner"] = $setup["idmap"]["owner01"];
        $commit_comment_ref01_data["repo"] = $setup["idmap"]["repo01"];

        $commit_comment_ref01_data_result = $commit_comment_ref01_ent->create($commit_comment_ref01_data, null);
        $commit_comment_ref01_data = Helpers::to_map(is_object($commit_comment_ref01_data_result) && method_exists($commit_comment_ref01_data_result, 'data_get') ? $commit_comment_ref01_data_result->data_get() : $commit_comment_ref01_data_result);
        $this->assertNotNull($commit_comment_ref01_data);
        $this->assertNotNull($commit_comment_ref01_data["id"]);

        // LIST
        $commit_comment_ref01_match = [
            "owner" => $setup["idmap"]["owner01"],
            "repo" => $setup["idmap"]["repo01"],
        ];

        $commit_comment_ref01_list_result = $commit_comment_ref01_ent->list($commit_comment_ref01_match, null);
        $this->assertIsArray($commit_comment_ref01_list_result);

        $found_item = sdk_select(
            Runner::entity_list_to_data($commit_comment_ref01_list_result),
            ["id" => $commit_comment_ref01_data["id"]]);
        $this->assertNotEmpty($found_item);

        // UPDATE
        $commit_comment_ref01_data_up0_up = [
            "id" => $commit_comment_ref01_data["id"],
            "owner" => $setup["idmap"]["owner"],
            "repo" => $setup["idmap"]["repo"],
        ];

        $commit_comment_ref01_markdef_up0_name = "author_association";
        $commit_comment_ref01_markdef_up0_value = "Mark01-commit_comment_ref01_" . $setup["now"];
        $commit_comment_ref01_data_up0_up[$commit_comment_ref01_markdef_up0_name] = $commit_comment_ref01_markdef_up0_value;

        $commit_comment_ref01_resdata_up0_result = $commit_comment_ref01_ent->update($commit_comment_ref01_data_up0_up, null);
        $commit_comment_ref01_resdata_up0 = Helpers::to_map(is_object($commit_comment_ref01_resdata_up0_result) && method_exists($commit_comment_ref01_resdata_up0_result, 'data_get') ? $commit_comment_ref01_resdata_up0_result->data_get() : $commit_comment_ref01_resdata_up0_result);
        $this->assertNotNull($commit_comment_ref01_resdata_up0);
        $this->assertEquals($commit_comment_ref01_resdata_up0["id"], $commit_comment_ref01_data_up0_up["id"]);
        $this->assertEquals($commit_comment_ref01_resdata_up0[$commit_comment_ref01_markdef_up0_name], $commit_comment_ref01_markdef_up0_value);

        // LOAD
        $commit_comment_ref01_match_dt0 = [
            "id" => $commit_comment_ref01_data["id"],
        ];
        $commit_comment_ref01_data_dt0_loaded = $commit_comment_ref01_ent->load($commit_comment_ref01_match_dt0, null);
        $commit_comment_ref01_data_dt0_load_result = Helpers::to_map(is_object($commit_comment_ref01_data_dt0_loaded) && method_exists($commit_comment_ref01_data_dt0_loaded, 'data_get') ? $commit_comment_ref01_data_dt0_loaded->data_get() : $commit_comment_ref01_data_dt0_loaded);
        $this->assertNotNull($commit_comment_ref01_data_dt0_load_result);
        $this->assertEquals($commit_comment_ref01_data_dt0_load_result["id"], $commit_comment_ref01_data["id"]);

    }
}

function commit_comment_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/commit_comment/CommitCommentTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = GithubSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["commit_comment01", "commit_comment02", "commit_comment03", "repo01", "repo02", "repo03", "commit01", "commit02", "commit03", "commit_sha01", "owner01"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("GITHUB_TEST_COMMIT_COMMENT_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "GITHUB_TEST_COMMIT_COMMENT_ENTID" => $idmap,
        "GITHUB_TEST_LIVE" => "FALSE",
        "GITHUB_TEST_EXPLAIN" => "FALSE",
        "GITHUB_APIKEY" => "",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["GITHUB_TEST_COMMIT_COMMENT_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }
    if (!isset($idmap_resolved["owner"])) {
        $idmap_resolved["owner"] = $idmap_resolved["owner01"];
    }
    if (!isset($idmap_resolved["repo"])) {
        $idmap_resolved["repo"] = $idmap_resolved["repo01"];
    }

    if ($env["GITHUB_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            // FIRST, so the generated fields below win: sdk-test-control.json's
            // test.client.options adds to the live client, it does not redirect it.
            Runner::live_client_options(),
            [
                "apikey" => $env["GITHUB_APIKEY"],
            ],
            // ismap, not a plain "?? []" default: an empty PHP array is a
            // LIST, and a non-map later entry REPLACES the accumulated map in
            // merge - so the no-extras call discarded live_client_options()
            // and the apikey/server map above it.
            Vs::ismap($extra) ? $extra : new \stdClass(),
        ]);
        $client = new GithubSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["GITHUB_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["GITHUB_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}
